client.go

```diff
@@ -219,6 +221,40 @@ func (co *Conn) ReadMsg() (*Msg, error) {
 	return m, err
 }
 
+// ReadMsgBytes reads a message bytes from the connection. Parses and fills
+// dns message wire header (passing nil would skip header parsing) and
+// returns message bytes to process them later.
+//
+// Note that this function would not be able to report TSIG error or
+// check it got actual DNS payload.
+func (co *Conn) ReadMsgBytes(hdr *Header) ([]byte, error) {
+	var p []byte
+	if _, ok := co.Conn.(*net.TCPConn); ok {
+		p = make([]byte, MaxMsgSize)
+	} else {
+		if co.UDPSize > MinMsgSize {
+			p = make([]byte, co.UDPSize)
+		} else {
+			p = make([]byte, MinMsgSize)
+		}
+	}
+
+	n, err := co.Read(p)
+	if err != nil {
+		return nil, err
+	} else if n < 12 {
+		return nil, ErrShortRead
+	}
+
+	p = p[:n]
+	if hdr != nil {
```

Comment:
It basically repeats Read but with allocating buffer for message bytes... Hm... Just because it's easier than creating new error about it I guess...